![Relax](https://raw.githubusercontent.com/interviewquery/takehomes/relax_1/relax_1/logo.png)

The data is available as two attached CSV files:
```
takehome_user_engagement.csv
takehome_users.csv
```
The data has the following two tables:

1.  A user table ( `takehome_users.csv`) with data on 12,000 users who
    signed up for the product in the last two years. This table
    includes:

    -   `name`: the user's name

    -   `object_id`: the user's id

    -   `email`: email address

    -   `creation_source`: how their account was created. This takes
        on one of 5 values:
        - `PERSONAL_PROJECTS`: invited to join another user's personal workspace
        - `GUEST_INVITE`: invited to an organization as a guest (limited permissions)
        - `ORG_INVITE`: invited to an organization (as a full member)
        - `SIGNUP`: signed up via the website
        - `SIGNUP_GOOGLE_AUTH`: signed up using Google Authentication (using a Google email account for their login id)

-   `creation_time`: when they created their account

-   `last_session_creation_time`: unix timestamp of last login

-   `opted_in_to_mailing_list`: whether they have opted into receiving
    marketing emails

-   `enabled_for_marketing_drip`: whether they are on the regular
    marketing email drip

-   `org_id`: the organization (group of users) they belong to

-   `invited_by_user_id`: which user invited them to join (if
    applicable).

2.  A usage summary table ( `takehome_user_engagement.csv` ) that has a
    row for each day that a user logged into the product.

Defining an _"adopted user"_ as a user who *has logged into the
product on three separate days in at least one seven-­day period*,
**identify which factors predict future user adoption**.

We suggest spending 1­2 hours on this, but you're welcome to spend more
or less.



In [4]:
!git clone --branch relax_1 https://github.com/interviewquery/takehomes.git
%cd takehomes/relax_1
!ls

Cloning into 'takehomes'...
remote: Enumerating objects: 1963, done.
remote: Counting objects: 100% (1963/1963), done.
remote: Compressing objects: 100% (1220/1220), done.
remote: Total 1963 (delta 752), reused 1928 (delta 726), pack-reused 0 (from 0)
Receiving objects: 100% (1963/1963), 297.43 MiB | 16.88 MiB/s, done.
Resolving deltas: 100% (752/752), done.
/content/takehomes/relax_1
example_solution  metadata.json       takehome_user_engagement.csv
logo.png	  takehomefile.ipynb  takehome_users.csv


In [ ]:
# fix the encodding to latian 1
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

okay redo we did a good job once again :

* what is an adiopted user: a user that visted / logined in 3 days in a week at least once each day  - caculate that specfic type (everything will be in comparsion to this metric on interest)
- calculte this number , get the rate for comparsions betwen the columns


0) this the portion where we discused cleaning the data to check the percentage of missing values and how to fix the missing values to make the data good and useful (oultiers or distribution type )
a) look at the measure of center
based on the data of what is an adopted user can we describe the data: such as the mean , median, avge of an adopted user (disributoin looking like )
b) look in the measuare of spread is it skwered are there oultileres, whars the sd variance in the data specfically of adopted user

c) we want to look at the different assosicaitons betwwen data poitns
- fitler and join the data apoprtiately (thinking about group by , etc )
- utlizing the columns to come up with hypothesis or associations checks

d) Hyphothesis:
- does a adoption rate have a higher rate in a specific create source than the other ?
(look at the segmentation accross the different users )


b does recieing maretking notfications have higher adoption user rate (either with markeiting drip or listing ) if so how

c) does the seasonality of time have higher rates of adoption inthe last two years if so is there a cycle or patten

e) After looking into the hyptoeses extracting interesting columns and looing into a corealtion heat map

f) running a feature selction aprroptatiley , preforming mulitilinear regeessions to see if columns of interes

g) look at the take home solution commet on the revisions of what we could improve on or fix, implement or unesscary





# step 1
read in both filed import pandas, numpy, matplot lib and seaborm

# step 2
look at one file at a tyime
get the head, describe the the columns and dtypes and column names,

# step 3
change neceary with a copy to the clean set
then visualize the import ance of missing data

start inderstand the spread and center of the time each year to get a sense of the distribution aka i think a density plot

# step 4
create the this adoption rate that made in the first data set  

In [20]:
users = pd.read_csv('takehome_users.csv', encoding = 'latin_1')
users.tail()

object_id        creation_time           name  \
11995      11996  2013-09-06 06:14:15   Meier Sophia   
11996      11997  2013-01-10 18:28:37  Fisher Amelie   
11997      11998  2014-04-27 12:45:16    Haynes Jake   
11998      11999  2012-05-31 11:55:59   Faber Annett   
11999      12000  2014-01-26 08:57:12     Lima Thaís   

                           email     creation_source  \
11995      SophiaMeier@gustr.com          ORG_INVITE   
11996     AmelieFisher@gmail.com  SIGNUP_GOOGLE_AUTH   
11997        JakeHaynes@cuvox.de        GUEST_INVITE   
11998         mhaerzxp@iuxiw.com   PERSONAL_PROJECTS   
11999  ThaisMeloLima@hotmail.com              SIGNUP   

       last_session_creation_time  opted_in_to_mailing_list  \
11995                1.378448e+09                         0   
11996                1.358275e+09                         0   
11997                1.398603e+09                         1   
11998                1.338638e+09                         0   
11999                1.390727e+09                         0   

       enabled_for_marketing_drip  org_id  invited_by_user_id  
11995                           0      89              8263.0  
11996                           0     200                 NaN  
11997                           1      83              8074.0  
11998                           0       6                 NaN  
11999                           1       0                 NaN

In [12]:
users.dtypes

object_id                       int64
creation_time                  object
name                           object
email                          object
creation_source                object
last_session_creation_time    float64
opted_in_to_mailing_list        int64
enabled_for_marketing_drip      int64
org_id                          int64
invited_by_user_id            float64
dtype: object

In [13]:
users.describe()

object_id  last_session_creation_time  opted_in_to_mailing_list  \
count  12000.00000                8.823000e+03              12000.000000   
mean    6000.50000                1.379279e+09                  0.249500   
std     3464.24595                1.953116e+07                  0.432742   
min        1.00000                1.338452e+09                  0.000000   
25%     3000.75000                1.363195e+09                  0.000000   
50%     6000.50000                1.382888e+09                  0.000000   
75%     9000.25000                1.398443e+09                  0.000000   
max    12000.00000                1.402067e+09                  1.000000   

       enabled_for_marketing_drip        org_id  invited_by_user_id  
count                12000.000000  12000.000000         6417.000000  
mean                     0.149333    141.884583         5962.957145  
std                      0.356432    124.056723         3383.761968  
min                      0.000000      0.000000            3.000000  
25%                      0.000000     29.000000         3058.000000  
50%                      0.000000    108.000000         5954.000000  
75%                      0.000000    238.250000         8817.000000  
max                      1.000000    416.000000        11999.000000

In [19]:
users['creation_time']= pd.to_datetime(users['creation_time'])
users.dtypes

object_id                              int64
creation_time                 datetime64[ns]
name                                  object
email                                 object
creation_source                       object
last_session_creation_time           float64
opted_in_to_mailing_list               int64
enabled_for_marketing_drip             int64
org_id                                 int64
invited_by_user_id                   float64
dtype: object

In [31]:
engagement = pd.read_csv('takehome_user_engagement.csv')
engagement.head()

time_stamp  user_id  visited
0  2014-04-22 03:53:30        1        1
1  2013-11-15 03:45:04        2        1
2  2013-11-29 03:45:04        2        1
3  2013-12-09 03:45:04        2        1
4  2013-12-25 03:45:04        2        1

In [23]:
engagement.dtypes

time_stamp    object
user_id        int64
visited        int64
dtype: object

In [25]:
engagement['time_stamp'] = pd.to_datetime(engagement['time_stamp'])

In [26]:
engagement.dtypes

time_stamp    datetime64[ns]
user_id                int64
visited                int64
dtype: object

In [42]:
for column in users:
  print(users[column].name, users[column].isnull().sum()/users[column].count())

object_id 0.0
creation_time 0.0
name 0.0
email 0.0
creation_source 0.0
last_session_creation_time 0.36008160489629376
opted_in_to_mailing_list 0.0
enabled_for_marketing_drip 0.0
org_id 0.0
invited_by_user_id 0.8700327255726975


In [58]:
# they booth seem to be clean
visited_boolean_check = engagement['visited'].unique()
print(visited_boolean_check )

users['opted_in_to_mailing_list'].unique()
users['enabled_for_marketing_drip'].unique()

# creation source should only have 3 pathways
users['creation_source'].unique()

# lets rename the object id column for more clarity in its place
users.rename(columns={'object_id':'user_id'})

[1]


user_id        creation_time               name  \
0            1  2014-04-22 03:53:30     Clausen August   
1            2  2013-11-15 03:45:04      Poole Matthew   
2            3  2013-03-19 23:14:52  Bottrill Mitchell   
3            4  2013-05-21 08:09:28    Clausen Nicklas   
4            5  2013-01-17 10:14:20          Raw Grace   
...        ...                  ...                ...   
11995    11996  2013-09-06 06:14:15       Meier Sophia   
11996    11997  2013-01-10 18:28:37      Fisher Amelie   
11997    11998  2014-04-27 12:45:16        Haynes Jake   
11998    11999  2012-05-31 11:55:59       Faber Annett   
11999    12000  2014-01-26 08:57:12         Lima Thaís   

                            email     creation_source  \
0        AugustCClausen@yahoo.com        GUEST_INVITE   
1          MatthewPoole@gustr.com          ORG_INVITE   
2      MitchellBottrill@gustr.com          ORG_INVITE   
3       NicklasSClausen@yahoo.com        GUEST_INVITE   
4              GraceRaw@yahoo.com        GUEST_INVITE   
...                           ...                 ...   
11995       SophiaMeier@gustr.com          ORG_INVITE   
11996      AmelieFisher@gmail.com  SIGNUP_GOOGLE_AUTH   
11997         JakeHaynes@cuvox.de        GUEST_INVITE   
11998          mhaerzxp@iuxiw.com   PERSONAL_PROJECTS   
11999   ThaisMeloLima@hotmail.com              SIGNUP   

       last_session_creation_time  opted_in_to_mailing_list  \
0                    1.398139e+09                         1   
1                    1.396238e+09                         0   
2                    1.363735e+09                         0   
3                    1.369210e+09                         0   
4                    1.358850e+09                         0   
...                           ...                       ...   
11995                1.378448e+09                         0   
11996                1.358275e+09                         0   
11997                1.398603e+09                         1   
11998                1.338638e+09                         0   
11999                1.390727e+09                         0   

       enabled_for_marketing_drip  org_id  invited_by_user_id  
0                               0      11             10803.0  
1                               0       1               316.0  
2                               0      94              1525.0  
3                               0       1              5151.0  
4                               0     193              5240.0  
...                           ...     ...                 ...  
11995                           0      89              8263.0  
11996                           0     200                 NaN  
11997                           1      83              8074.0  
11998                           0       6                 NaN  
11999                           1       0                 NaN  

[12000 rows x 10 columns]

- we can see the invited by user id has an 87 % of null values , make sure there filled as 0 if no user id is present
-  we can also the last created time also has about 36 % of null values and also 0 for the las time , is it a time stamp


 - we see that in the users file there is no nulls ,
one thing to check is if the boolean values are true,
change the orject id to user id ? i beleive there the users confusing tiitle name ,

# next lets see if there is any erros where a user id invited their self ? ,
# last created session is a time stame needs to be converted
# fill in the large gap of nulls with 0 if its appropriate